adapters/ads_adapter.py

```diff
@@ -1,22 +1,173 @@
-""" Beckhoff ADS adapter.
+"""Beckhoff ADS adapter.
 
-Downloads all beckhof ADS data and sends it to the GUI tool.
+Subscribes to Beckhof ADS variables and sends it to the GUI tool.
+
+Auto subscribe only works for DINT and LREAL for now.
+
+Example usage:
+    python ads_adapter.py --regex GVl.*temperature
 
 """
+from lognplot.client import LognplotTcpClient
+import pyads
+import time
+import ctypes
+import struct
+import datetime
+import re
+import argparse
+from collections import namedtuple
+from typing import Type
 
-# TODO: implement.
 
+class AdsClient:
+    Entry = namedtuple(
+        "Entry", ["name", "typename", "comment", "datatype", "datatype_size"]
+    )
 
+    DATATYPE_MAP = {
+        "BOOL": pyads.PLCTYPE_BOOL,
+        "BYTE": pyads.PLCTYPE_BYTE,
+        "DINT": pyads.PLCTYPE_DINT,
+        "DWORD": pyads.PLCTYPE_DWORD,
+        "INT": pyads.PLCTYPE_INT,
+        "LREAL": pyads.PLCTYPE_LREAL,
+        "REAL": pyads.PLCTYPE_REAL,
+        "SINT": pyads.PLCTYPE_SINT,
+        "UDINT": pyads.PLCTYPE_UDINT,
+        "UINT": pyads.PLCTYPE_UINT,
+        "USINT": pyads.PLCTYPE_USINT,
+        "WORD": pyads.PLCTYPE_WORD,
+    }  # type: Dict[str, Type]
 
-from lognplot.client import LognplotTcpClient
+    def __init__(self, ams_net_id: str, ams_net_port: str):
+        self.notification_handles = []
+        self.plc = pyads.Connection(ams_net_id, ams_net_port)
+        self.plc.open()
+
+    def __del__(self):
+        for handle in self.notification_handles:
+            self.plc.del_device_notification(*handle)
+
+        self.plc.close()
+
+    def log(self, name: str, plc_type):
```

Comment:
I was a bit confused by this name, since I thought it would log a logmessage. Maybe rename this to `subscribe_to_change_notifications` or something like that?